drivers/platform/stm32/stm32_gpio_irq.c

```diff
@@ -0,0 +1,497 @@
+	if (ret != HAL_OK)
+		return -EFAULT;
+
+	switch (level) {
+	case NO_OS_IRQ_EDGE_FALLING:
+		config.Trigger = EXTI_TRIGGER_FALLING;
+		break;
+	case NO_OS_IRQ_EDGE_RISING:
+		config.Trigger = EXTI_TRIGGER_RISING;
+		break;
+	case NO_OS_IRQ_EDGE_BOTH:
+		config.Trigger = EXTI_TRIGGER_RISING_FALLING;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	ret = HAL_EXTI_SetConfigLine(&sdesc->hexti, &config);
+	if (ret != HAL_OK)
+		return -EFAULT;
+
+	return 0;
+}
+
+/**
+ * @brief Register a callback function to be triggered when an
+ * interrupt occurs.
+ * @param desc - The GPIO IRQ controller descriptor.
+ * @param irq_id - the pin on which the interrupt signal will be.
+ * @param cb - Descriptor of the callback.
+ * @return 0 if successfull, negative error code otherwise.
+ */
+static int32_t stm32_gpio_irq_register_callback(struct no_os_irq_ctrl_desc
+		*desc,
+		uint32_t irq_id,
+		struct no_os_callback_desc *cb)
+{
+	int ret;
+	struct irq_action *action;
+	struct irq_action action_key = {.irq_id = irq_id};
+	struct stm32_gpio_irq_desc *sdesc;
+
+	if (!desc || !desc->extra || !cb || !IS_EXTI_GPIO_PIN(irq_id))
+		return -EINVAL;
+
+	sdesc = desc->extra;
+
+	ret = no_os_list_read_find(actions, (void **)&action, &action_key);
+	/*
+	* If no action was found, insert a new one, otherwise update it
+	*/
+	if (ret) {
+		action = calloc(1, sizeof(*action));
+		if (!action)
+			return -ENOMEM;
+
+		action->irq_id = irq_id;
+		action->ctx = cb->ctx;
+		action->callback = cb->callback;
+
+		ret = no_os_list_add_last(actions, action);
+		if (ret)
+			goto free_action;
+	} else {
+		action->irq_id = irq_id;
+		action->ctx = cb->ctx;
+		action->callback = cb->callback;
+	}
+
+	EXTI_ConfigTypeDef config;
+	config.Mode = EXTI_MODE_INTERRUPT;
+	config.GPIOSel = sdesc->port_nb;
+	config.Line = EXTI_GPIO | irq_id;
+	ret = HAL_EXTI_SetConfigLine(&sdesc->hexti, &config);
+	if (ret) {
+		ret = -EFAULT;
+		goto free_action;
+	}
+
+	return 0;
+
+free_action:
+	free(action);
+	return ret;
+}
+
+/**
+ * @brief Unregister a callback.
+ * @param desc - GPIO interrupt controller descriptor.
+ * @param irq_id - Id of the interrupt
+ * @param cb - Descriptor of the callback.
+ * @return 0 if successfull, negative error code otherwise.
+ */
+static int32_t stm32_gpio_irq_unregister_callback(struct no_os_irq_ctrl_desc
+		*desc,
+		uint32_t irq_id, struct no_os_callback_desc *cb)
+{
+	int ret;
+	struct irq_action *discard_action = NULL;
+	struct irq_action action_key = {.irq_id = irq_id};
+
+	if (!desc || !cb || !IS_EXTI_GPIO_PIN(irq_id))
+		return -EINVAL;
+
+	ret = no_os_list_read_find(actions, (void **)&discard_action, &action_key);
+	if (ret)
+		return -ENODEV;
+
+	free(discard_action);
+	return 0;
+}
+
+/**
+ * @brief Enable all GPIO interrupts.
+ * @param desc - the GPIO irq descriptor.
+ * @return 0
+ */
+static int32_t stm32_gpio_irq_global_enable(struct no_os_irq_ctrl_desc *desc)
+{
+	/* Enable all available EXTI line interrupts */
+	HAL_NVIC_EnableIRQ(EXTI0_IRQn);
```

Comment:
and if you implement the desc->parent part, you may simply do no_os_gpio_irq_global_enable(desc->parent)

Reply:
In this case stm32 gpio controller does not allow to globally disable or enable all gpio interrupts so I decided to simply return -ENOSYS. However, for example on maxim platform, the gpio irq controller allows to disable/enable all gpio interrupts. So the general idea is that gpio_irq_global_enable/disable should be used to enable/disable all gpio specifc interrupts if possible, otherwise it should return -ENOSYS. On the other hand if a user wants to enable/disable global interrupts he should use the api from NVIC controller directly.